application/src/test/java/org/thingsboard/server/controller/AbstractWebTest.java

```diff
@@ -330,18 +333,39 @@ public void teardownWebTest() throws Exception {
         log.debug("Executing web test teardown");
 
         loginSysAdmin();
-        doDelete("/api/tenant/" + tenantId.getId().toString())
-                .andExpect(status().isOk());
+        deleteTenant(tenantId);
         deleteDifferentTenant();
-
         verifyNoTenantsLeft();
 
         tenantProfileService.deleteTenantProfiles(TenantId.SYS_TENANT_ID);
 
         log.info("Executed web test teardown");
     }
 
-    void verifyNoTenantsLeft() throws Exception {
+    private void verifyNoTenantsLeft() throws Exception {
+        List<Tenant> loadedTenants = getAllTenants();
+        if (!loadedTenants.isEmpty()) {
+            loadedTenants.forEach(tenant -> deleteTenant(tenant.getId()));
+            loadedTenants = getAllTenants();
+        }
+        assertThat(loadedTenants).as("All tenants expected to be deleted, but some tenants left in the database").isEmpty();
+    }
+
+    private void deleteTenant(TenantId tenantId) {
+        int status = 0;
+        int retries = 0;
+        while (status != HttpStatus.SC_OK && retries < CLEANUP_TENANT_RETRIES_COUNT) {
```

Comment:
In general, the problem is that the tenant can not be deleted is something permanent. like some related entity having a foreign key, etc.
It might be helpful for some cases, when the test fails and not delete additional tenants. Will see...

Reply:
Hey, thank you for the review! This fix is dedicated to the race condition when some unprocessed messages fail tenant deletion. The plan is to create test tenants only once before all tests and not do clean-up (it could save some build time)